File: 2015/day3.jl
data = open("./input/day3.txt") do file
    read(file, String)
end

#  Part 1
function part_one(d)
    seen = Set()
    pos = [0,0]
    push!(seen, pos)

    for char in split(d, "")
        if char == "^"
            pos += [-1,0]
            push!(seen, pos)
        elseif char == ">"
            pos += [0,1]
            push!(seen, pos)
        elseif char == "<"
            pos += [0,-1]
            push!(seen, pos)
        elseif char == "v"
            pos += [1,0]
            push!(seen, pos)
        end
    end
    return seen
end
println(length(part_one(data)))

# Part 2
function part_two()
    seen = Set()
    santa = [0,0]
    robo_santa = [0,0]
    push!(seen, santa)

    santa_turn = true
    for char in split(data, "")
        if santa_turn
            if char == "^"
                santa += [-1,0]
                push!(seen, santa)
            elseif char == ">"
                santa += [0,1]
                push!(seen, santa)
            elseif char == "<"
                santa += [0,-1]
                push!(seen, santa)
            elseif char == "v"
                santa += [1,0]
                push!(seen, santa)
            end
            santa_turn = false
        else
            if char == "^"
                robo_santa += [-1,0]
                push!(seen, robo_santa)
            elseif char == ">"
                robo_santa += [0,1]
                push!(seen, robo_santa)
            elseif char == "<"
                robo_santa += [0,-1]
                push!(seen, robo_santa)
            elseif char == "v"
                robo_santa += [1,0]
                push!(seen, robo_santa)
            end
            santa_turn = true
        end
    end
    return seen
end
println(length(part_two()))

File: 2024/day3.py
import re

data = ""
with open("input/day3.txt", "r") as f:
    lines = f.readlines()
    for line in lines:
        data += line

valid_instructions = re.findall(r"mul\(\d{1,3},\d{1,3}\)|do\(\)|don\'t\(\)", data)

# part 1
products1 = []
for mul in valid_instructions:
    try:
        nums = mul[4:-1]
        f1, f2 = nums.split(",")
        products1.append(int(f1) * int(f2))
    except ValueError:
        continue

print(sum(products1))

# part 2
products2 = []
process = True
for inst in valid_instructions:
    if inst == "do()":
        process = True
    elif inst == "don't()":
        process = False
    elif process:
        nums = inst[4:-1]
        f1, f2 = nums.split(",")
        products2.append(int(f1) * int(f2))

print(sum(products2))
